models/lowpassfilter_spike_detector.cpp

```diff
@@ -0,0 +1,856 @@
+
+  std::vector< std::vector< double > > tmp_node_traces(
+    S_.node_gids_.size(), std::vector< double >() );
+  S_.node_traces_.swap( tmp_node_traces );
+
+  // Size is known.
+  std::vector< double > tmp_trace_times( S_.node_gids_.size(), 0.0 );
+  B_.trace_times_.swap( tmp_trace_times );
+
+  // Size is known.
+  std::vector< double > tmp_traces( S_.node_gids_.size(), 0.0 );
+  B_.traces_.swap( tmp_traces );
+
+  std::vector< long > tmp_steps_to_filter( 0 );
+  B_.steps_to_filter_.swap( tmp_steps_to_filter );
+
+  std::vector< std::vector< std::vector< Event* > > > tmp_node_spikes(
+    2,
+    std::vector< std::vector< Event* > >(
+      S_.node_gids_.size(), std::vector< Event* >() ) );
+  B_.node_spikes_.swap( tmp_node_spikes );
+}
+
+void
+nest::lowpassfilter_spike_detector::calibrate()
+{
+  if ( kernel().event_delivery_manager.get_off_grid_communication()
+    and not spikes_device_.is_precise_times_user_set() )
+  {
+    spikes_device_.set_precise_times( true );
+    filtered_device_.set_precise_times( true );
+    std::string msg = String::compose(
+      "Precise neuron models exist: the property precise_times "
+      "of the %1 with gid %2 has been set to true",
+      get_name(),
+      get_gid() );
+
+    if ( spikes_device_.is_precision_user_set() )
+    {
+      // if user explicitly set the precision, there is no need to do anything.
+      msg += ".";
+    }
+
+    else
+    {
+      // it makes sense to increase the precision if precise models are used.
+      spikes_device_.set_precision( 15 );
+      filtered_device_.set_precision( 15 );
+      msg += ", precision has been set to 15.";
+    }
+
+    LOG( M_INFO, "spike_detector::calibrate", msg );
+  }
+
+  spikes_device_.calibrate();
+  filtered_device_.calibrate();
+}
+
+void
+nest::lowpassfilter_spike_detector::print_value( long sender,
+  double time,
+  double value )
+{
+  filtered_device_.print_value( sender, false );
+  filtered_device_.print_value( time, false );
+  filtered_device_.print_value( value, true );
+}
+
+long
+nest::lowpassfilter_spike_detector::filter_step_( long update_start )
+{
+  // This method returns the first step to report based on the current filter
+  // block.
+
+  long interval_step = P_.filter_report_interval_.get_steps();
+  long filterblock_start_step =
+    Time( Time::ms( P_.filter_start_times_[ V_.filter_block_index_ ] ) )
+      .get_steps();
+  long filter_step;
+
+  if ( filterblock_start_step + interval_step > update_start )
+  {
+    filter_step = filterblock_start_step + interval_step;
+  }
+
+  else
+  {
+    filter_step = filterblock_start_step + interval_step
+      + std::ceil( ( ( double ) ( ( update_start + 1 )
+                       - ( filterblock_start_step + interval_step ) )
+          / interval_step ) ) * interval_step;
+
+    if ( ( filter_step - ( filterblock_start_step + interval_step ) )
+        % interval_step
+      != 0 )
+    {
+      filter_step = filter_step - ( filter_step % interval_step );
+    }
```

Comment:
Addressed. This is tested and works fast. In a separate pull request, attempts can be made to incorporate 2.2 into 2.1 or replace 2. with for loops (which have drawbacks compared to this and may or may not seem simpler).